### PR TITLE
Build with Java 21

### PR DESCRIPTION
**Pull request by groldan**
Updates to make the build pass with Java 21.

```bash
java -version
openjdk version "21" 2023-09-19 LTS
OpenJDK Runtime Environment Temurin-21+35 (build 21+35-LTS)
OpenJDK 64-Bit Server VM Temurin-21+35 (build 21+35-LTS, mixed mode, sharing)
```

So far only the core modules build, haven't run `-Prelease,communityRelease`

  

# Checklist

- [ ] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [ ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [ ] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by groldan** [comment]
`mvn test -pl :gs-wms` works, but when running all projects tests I get an " `java.lang.VirtualMachineError: Out of space in CodeCache for adapters`" error:

```
[ERROR] Tests run: 88, Failures: 10, Errors: 3, Skipped: 0, Time elapsed: 2.637 s <<< FAILURE! - in org.geoserver.wms.wms_1_1_1.GetMapIntegrationTest
[ERROR] testMosaicHoles(org.geoserver.wms.wms_1_1_1.GetMapIntegrationTest)  Time elapsed: 0.266 s  <<< ERROR!
java.io.IOException: org.geotools.api.data.DataSourceException: Unable to create this mosaic
	at org.geoserver.wms.wms_1_1_1.GetMapIntegrationTest.onSetUp(GetMapIntegrationTest.java:281)
Caused by: org.geotools.api.data.DataSourceException: Unable to create this mosaic
	at org.geoserver.wms.wms_1_1_1.GetMapIntegrationTest.onSetUp(GetMapIntegrationTest.java:281)
Caused by: java.io.IOException: java.util.concurrent.ExecutionException: org.geotools.gce.imagemosaic.GranuleLoadingException: Failed to load granule file:/home/groldan/git/geoserver/geoserver/master/src/wms/./target/default1175587496092344886data/mosaicTazDem/tazdemFloat0.tif
	at org.geoserver.wms.wms_1_1_1.GetMapIntegrationTest.onSetUp(GetMapIntegrationTest.java:281)
Caused by: java.util.concurrent.ExecutionException: org.geotools.gce.imagemosaic.GranuleLoadingException: Failed to load granule file:/home/groldan/git/geoserver/geoserver/master/src/wms/./target/default1175587496092344886data/mosaicTazDem/tazdemFloat0.tif
	at org.geoserver.wms.wms_1_1_1.GetMapIntegrationTest.onSetUp(GetMapIntegrationTest.java:281)
Caused by: org.geotools.gce.imagemosaic.GranuleLoadingException: Failed to load granule file:/home/groldan/git/geoserver/geoserver/master/src/wms/./target/default1175587496092344886data/mosaicTazDem/tazdemFloat0.tif
	at org.geoserver.wms.wms_1_1_1.GetMapIntegrationTest.onSetUp(GetMapIntegrationTest.java:281)
Caused by: javax.media.jai.util.ImagingException: All factories fail for the operation "Scale"
	at org.geoserver.wms.wms_1_1_1.GetMapIntegrationTest.onSetUp(GetMapIntegrationTest.java:281)
Caused by: java.lang.reflect.InvocationTargetException
	at org.geoserver.wms.wms_1_1_1.GetMapIntegrationTest.onSetUp(GetMapIntegrationTest.java:281)
Caused by: java.lang.VirtualMachineError: Out of space in CodeCache for adapters
	at org.geoserver.wms.wms_1_1_1.GetMapIntegrationTest.onSetUp(GetMapIntegrationTest.java:281)
```

**Comment by aaime** [comment]
The error is really weird... The JVMs are reused for tests, so I guess it's due to an accumulation of sorts. Not sure why it's happening only with Java 21 though? There is a system variable that we can set, but not sure how well that work... if it build up, I guess we'll end up having the same issue later in the build:

https://knowledge.informatica.com/s/article/488211?language=en_US

**Comment by groldan** [comment]
Looks like the JVM ergonomics are setting the CodeCache size too small due to the `-XX:+TieredCompilation` in the pom's `argLine`, which sets `-Xmx512m` and `-XX:+TieredCompilation`, with CodeCache size resuling in ~50M:

```
java -Xmx512m -XX:+UnlockDiagnosticVMOptions -XX:+PrintFlagsFinal -XX:+TieredCompilation -XX:TieredStopAtLevel=1 -version | grep -i reserved
    uintx ReservedCodeCacheSize                    = 50331648                               {pd product} {default}
     bool RestrictReservedStack                    = true                                      {product} {default}
     intx StackReservedPages                       = 1                                      {pd product} {default}
openjdk version "21" 2023-09-19 LTS
OpenJDK Runtime Environment Temurin-21+35 (build 21+35-LTS)
OpenJDK 64-Bit Server VM Temurin-21+35 (build 21+35-LTS, mixed mode, emulated-client, sharing)
```

Setting it explicitly to 64M through `-XX:ReservedCodeCacheSize=64m` in `argLine` seems to fix it,  though I wonder if it's too close to the limit:

```
java -Xmx512m -XX:+UnlockDiagnosticVMOptions -XX:+PrintFlagsFinal -XX:ReservedCodeCacheSize=64m -XX:+TieredCompilation -XX:TieredStopAtLevel=1 -version | grep -i reserved
    uintx ReservedCodeCacheSize                    = 67108864                               {pd product} {command line}
     bool RestrictReservedStack                    = true                                      {product} {default}
     intx StackReservedPages                       = 1                                      {pd product} {default}
openjdk version "21" 2023-09-19 LTS
OpenJDK Runtime Environment Temurin-21+35 (build 21+35-LTS)
OpenJDK 64-Bit Server VM Temurin-21+35 (build 21+35-LTS, mixed mode, emulated-client, sharing)
```

**Comment by groldan** [comment]
yes it very well could be an accumulation. I'm trying replacing `<forkMode>once</forkMode>` by `<reuseForks>false</reuseForks> for the surefire plugin. `forkMode` is deprecated, `reuseForks` *"Indicates if forked VMs can be reused. If set to "false", a new VM is forked for each test class to be executed. If set to "true", up to forkCount VMs will be forked and then reused to execute all tests."*

**Comment by groldan** [comment]
So far with `reulseForks = false` I'm getting a couple test failures but not the codecache error anymore.
One failure is
```
 testGetAsVersionsJSON(org.geoserver.rest.catalog.AboutControllerTest)  Time elapsed: 0.005 s  <<< FAILURE!
java.lang.AssertionError
	at org.geoserver.rest.catalog.AboutControllerTest.checkJSONModel(AboutControllerTest.java:142)
	at org.geoserver.rest.catalog.AboutControllerTest.testGetAsVersionsJSON(AboutControllerTest.java:81)
```
Probably unrelated and due to git commitd plugin not supporting git worktrees (I've to run the build with `-Dmaven.gitcommitid.nativegit=true`)

The other is 10 out of 14 tests in `org.geoserver.wcs2_0.kvp.ScaleKvpTest`, e.g.
```

[ERROR] scaleToSize[1](org.geoserver.wcs2_0.kvp.ScaleKvpTest)  Time elapsed: 0.002 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[image/tiff]> but was:<[application/xml]>
	at org.geoserver.wcs2_0.kvp.ScaleKvpTest.scaleToSize(ScaleKvpTest.java:344)

[ERROR] scaleAxesByFactor[1](org.geoserver.wcs2_0.kvp.ScaleKvpTest)  Time elapsed: 0.014 s  <<< FAILURE!
java.lang.AssertionError: 
expected:<GEOGCS["WGS 84", 
  DATUM["World Geodetic System 1984"
```

**Comment by jodygarnett** [comment]
@groldan if you rebase you should be able to uncomment the Java 21 .github/workflow/linux.yml lines:

https://github.com/geoserver/geoserver/blob/ec321c431f8797c879e4cb47ec1307115676c5b7/.github/workflows/linux.yml#L20-L26

This line is interesting as it allows jdk 21 build to run, but not fail the entire build.
```
continue-on-error: ${{ matrix.experimental }}
```

**Comment by aaime** [comment]
The reason for the tiered compilation flag, was that tests did run faster that way: they are relatively short-lived JVMs where making a big compile effort was not going to pay off. However, that was done 7 years ago, things might have changed with more recent VMs. 

Here are some preliminary findings:

- Tiered compilation is apparently on by default since Java 8, the actual flag changing things seems to be ``-XX:TieredStopAtLevel=1``, which stops native compilation at C1 (rather than the more expensive C2)
- Java 21.0 fails for me regardless of whether the Tiered stop at level 1is used or not, but I get a hard JVM crash rather than a code cache error (which I also get, when using tiered compilation)
- Java 21.0.1 is out and does complete the build if the tiered compilation flags are removed
- Java 21 without the stop at level 1 dedicates 251658240 bytes (240m) to the ReservedCodeCacheSize, so I guess that would be a good value to play with.

Also, here are some build timings, in seconds, of `` :

| JVM         | Tiered            | No tiered flags       |
| ----------- | ----------------- | --------------------- |
| Java11      | 528               | 643                   |
| Java 17     | 512               | 618                   |
| Java 21.0   | Fail (code cache) | Fail (hard JVM crash) |
| Java 21.0.1 | Fail (code cache) | 540                   |

But also, Java 21 with ``-XX:ReservedCodeCacheSize=251658240`` --> 507 seconds. 
Tested that flag with Java 11, it works there too.

Long story short: keep the tiered compilation, it generally helps, and just make the reserved code cache be 240m like in a normal JVM.
